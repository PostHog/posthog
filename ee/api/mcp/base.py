from mcp_server.djangomcp import DjangoMCP

mcp = DjangoMCP(name="PostHog MCP", instructions="Query analytics, errors and manage feature flags.")


@mcp.tool()
def add(a: int, b: int) -> int:
    return a + b
