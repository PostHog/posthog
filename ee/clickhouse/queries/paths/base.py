from abc import ABC
from typing import List, Tuple

from ee.clickhouse.client import sync_execute
from ee.clickhouse.queries.paths.path_event_query import PathEventQuery
from ee.clickhouse.sql.paths.path import PATH_ARRAY_QUERY
from posthog.constants import FUNNEL_WINDOW_INTERVAL, FUNNEL_WINDOW_INTERVAL_UNIT, LIMIT
from posthog.models import Filter, Team
from posthog.models.filters.path_filter import PathFilter
from posthog.queries.funnel import Funnel

EVENT_IN_SESSION_LIMIT_DEFAULT = 5
SESSION_TIME_THRESHOLD_DEFAULT = 1800000  # milliseconds to 30 minutes


class ClickhousePathBase(ABC, Funnel):
    _filter: PathFilter
    _team: Team

    def __init__(self, filter: PathFilter, team: Team) -> None:
        self._filter = filter
        self._team = team
        self.params = {
            "team_id": self._team.pk,
            "events": [],  # purely a speed optimization, don't need this for filtering
            "event_in_session_limit": EVENT_IN_SESSION_LIMIT_DEFAULT,
            "session_time_threshold": SESSION_TIME_THRESHOLD_DEFAULT,
            "autocapture_match": "%autocapture:%",
        }

    def run(self, *args, **kwargs):

        results = self._exec_query()
        return self._format_results(results)

    def _format_results(self, results):
        if not results or len(results) == 0:
            return []

        resp = []
        for res in results:
            resp.append(
                {"source": res[0], "target": res[1], "value": res[2],}
            )
        return resp

    def _exec_query(self) -> List[Tuple]:
        query = self.get_query()
        return sync_execute(query, self.params)

    def get_query(self) -> str:
        path_event_query, params = PathEventQuery(filter=self._filter, team_id=self._team.pk).get_query()
        self.params.update(params)
        return PATH_ARRAY_QUERY.format(path_event_query=path_event_query, boundary_event_filter="")
