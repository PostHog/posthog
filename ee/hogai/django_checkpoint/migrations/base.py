import logging
from abc import ABC, abstractmethod
from datetime import datetime, UTC
from typing import Any

from ee.hogai.utils.types import GraphContext, GraphType, VersionMetadata
from .registry import registry

from ee.models.assistant import ConversationCheckpoint, ConversationCheckpointBlob, ConversationCheckpointWrite

logger = logging.getLogger(__name__)


class BaseMigration(ABC):
    """
    Abstract base class for all checkpoint migrations.

    This class provides common functionality for:
    - Blob and write I/O operations
    - Checkpoint metadata management
    - Migration orchestration

    Subclasses must implement:
    - needs_migration(): Detection logic for this migration
    - apply_to_state_object(): State transformation logic
    """

    @classmethod
    def get_version(cls) -> int:
        """
        Get the version number for this migration class.

        Returns:
            Version number for this migration
        """
        return registry.get_version_for_class(cls)

    @staticmethod
    @abstractmethod
    def needs_migration(state_obj: Any) -> bool:
        """
        Check if a state object needs this specific migration.

        This is migration-specific logic that determines whether
        this migration should be applied to a given state object.

        Args:
            state_obj: Decoded state object from checkpoint blob or metadata

        Returns:
            True if this migration is needed, False otherwise
        """
        pass

    @staticmethod
    @abstractmethod
    def apply_to_state_object(state_obj: Any, graph_type: GraphType, context: GraphContext) -> Any:
        """
        Apply this migration to a state object.

        This is the core transformation logic that each migration
        implements to transform state objects.

        Args:
            state_obj: The state object to migrate
            graph_type: Graph type for migration context
            context: Context for migration context

        Returns:
            Migrated state object
        """
        pass

    @classmethod
    async def apply_to_blob_or_write(
        cls,
        object: ConversationCheckpointBlob | ConversationCheckpointWrite,
        checkpointer_serde,
        graph_type: GraphType,
        context: GraphContext,
    ) -> bool:
        """
        Apply this migration to a checkpoint blob and save it back to database.

        Args:
            object: ConversationCheckpointBlob or ConversationCheckpointWrite to migrate
            checkpointer_serde: Serializer from checkpointer
            graph_type: Graph type for migration context
            context: Context for migration context

        Returns:
            True if migration was applied and saved, False if no migration needed
        """
        # Skip empty or null blobs
        if not object.blob or object.type in ["empty", "null", None]:
            return False

        try:
            # Decode the blob data
            decoded_data = checkpointer_serde.loads_typed((object.type, object.blob))

            # Check if migration is needed (migration-specific logic)
            if not cls.needs_migration(decoded_data):
                return False

            # Apply migration (migration-specific logic)
            migrated_data = cls.apply_to_state_object(decoded_data, graph_type, context)

            # Re-encode the migrated data
            new_type, new_blob = checkpointer_serde.dumps_typed(migrated_data)

            # Save back to database
            object.type = new_type
            object.blob = new_blob
            await object.asave(update_fields=["type", "blob"])

            logger.info(
                f"{cls.__name__}: Successfully migrated and saved checkpoint blob {object.id} "
                f"({object.channel}, {context}, v{cls.get_version()})"
            )

            return True

        except Exception as e:
            logger.exception(f"Failed to migrate checkpoint blob {object.id}, error: {e}")
            return False

    @classmethod
    async def apply_to_checkpoint(
        cls, checkpoint: ConversationCheckpoint, checkpointer_serde, graph_type: GraphType, context: GraphContext
    ) -> bool:
        """
        Apply this migration to an entire checkpoint (metadata + blobs + writes).

        This orchestrates the migration of a complete checkpoint by:
        1. Checking if migration is already applied
        2. Detecting graph context
        3. Migrating blobs and writes
        4. Updating checkpoint metadata

        Args:
            checkpoint: ConversationCheckpoint to migrate
            checkpointer_serde: Serializer from checkpointer

        Returns:
            Tuple of (migration_applied, migration_stats)
        """
        # Check if checkpoint is already migrated to this version or higher
        metadata = checkpoint.metadata or {}
        version_info = metadata.get("version_metadata", {})
        current_schema_version = version_info.get("schema_version", 0)

        if current_schema_version >= cls.get_version():
            return False

        if hasattr(cls, "detect_graph_context"):
            # Specific to migration 0001, we overwrite graph_type and context
            # detecting them from the checkpoint
            graph_type, context = cls.detect_graph_context(checkpoint)

        migration_applied = False

        # Migrate all blobs in this checkpoint
        async for blob in checkpoint.blobs.all():
            if await cls.apply_to_blob_or_write(blob, checkpointer_serde, graph_type, context):
                migration_applied = True

        # Migrate all writes in this checkpoint
        async for write in checkpoint.writes.all():
            if await cls.apply_to_blob_or_write(write, checkpointer_serde, graph_type, context):
                migration_applied = True

        # Update checkpoint metadata to mark this migration as complete
        version = cls.get_version()
        if migration_applied or current_schema_version < version:
            updated_metadata = metadata.copy()
            updated_metadata["version_metadata"] = VersionMetadata(
                schema_version=version,
                migrated_at=datetime.now(UTC).isoformat(),
                graph_type=graph_type,
                context=context,
            ).model_dump(mode="json")

            checkpoint.metadata = updated_metadata
            await checkpoint.asave(update_fields=["metadata"])

        return migration_applied
