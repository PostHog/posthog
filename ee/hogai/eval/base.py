import os
import asyncio
from collections.abc import Sequence
from functools import partial

import pytest

from braintrust import EvalAsync, EvalCase, Metadata, init_logger
from braintrust.framework import EvalData, EvalScorer, EvalTask, Input, Output

from posthog.models.utils import uuid7

from ee.hogai.eval.schema import DatasetInput


async def _filter_data(data: EvalData[Input, Output], case_filter: str | None = None):
    # Resolve async data
    if asyncio.iscoroutine(data):
        # Async iterator
        if hasattr(data, "__aiter__"):
            data = [case async for case in data]
        # asyncio.iscoroutine may return True for sync generators
        elif hasattr(data, "__iter__"):
            data = list(data)
        # Regular awaitable
        else:
            data = await data
    cases = []
    for case in data:  # type: ignore
        if not isinstance(case, EvalCase):
            cases.append(case)
            continue

        # Reset trace IDs for DatasetInput, so we use distinct IDs instead
        if os.getenv("EVAL_MODE") == "offline" and isinstance(case.input, DatasetInput):
            # Mutating the input in place is intentional here. Just avoiding copying the whole object.
            case.input.trace_id = str(uuid7())

        # Filter by --case <eval_case_name_part> pytest flag
        if case_filter:
            if case_filter in str(case.input):
                cases.append(case)
        else:
            cases.append(case)

    return cases


async def BaseMaxEval(
    experiment_name: str,
    data: EvalData[Input, Output],
    task: EvalTask[Input, Output],
    scores: Sequence[EvalScorer[Input, Output]],
    pytestconfig: pytest.Config,
    metadata: Metadata | None = None,
    is_public: bool = False,
    no_send_logs: bool = True,
):
    if is_public and not no_send_logs:
        # We need to specify a separate project for each MaxEval() suite for comparison to baseline to work
        # That's the way Braintrust folks recommended - Braintrust projects are much more lightweight than PostHog ones
        project_name = f"max-ai-{experiment_name}"
        init_logger(project_name)
    else:
        project_name = experiment_name

    case_filter = pytestconfig.option.eval

    timeout = 60 * 8  # 8 minutes
    if os.getenv("EVAL_MODE") == "offline":
        timeout = 60 * 60  # 1 hour

    result = await EvalAsync(
        project_name,
        data=await _filter_data(data, case_filter=case_filter),
        task=task,
        scores=scores,
        timeout=timeout,
        max_concurrency=100,
        is_public=is_public,
        no_send_logs=no_send_logs,
        metadata=metadata,
    )

    # If we're running in the offline mode and the test case marked as public, the pipeline must completely fail.
    if os.getenv("EVAL_MODE") == "offline" and is_public:
        raise RuntimeError("Evaluation cases must be private when EVAL_MODE is set to offline.")

    if os.getenv("EXPORT_EVAL_RESULTS"):
        with open("eval_results.jsonl", "a") as f:
            f.write(result.summary.as_json() + "\n")

    return result


MaxPublicEval = partial(BaseMaxEval, is_public=True, no_send_logs=False)
"""Evaluation case that is publicly accessible."""

MaxPrivateEval = partial(BaseMaxEval, is_public=False, no_send_logs=True)
"""Evaluation case is not accessible publicly."""
