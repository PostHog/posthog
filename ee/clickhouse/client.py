import asyncio
import hashlib
import json
import types
from time import perf_counter
from typing import Any, Dict, List, Optional, Tuple, Union, cast

import sqlparse
from aioch import Client
from asgiref.sync import async_to_sync
from clickhouse_driver import Client as SyncClient
from clickhouse_pool import ChPool
from django.conf import settings as app_settings
from django.core.cache import cache
from django.utils.timezone import now
from sentry_sdk.api import capture_exception

from ee.clickhouse.errors import wrap_query_error
from ee.clickhouse.timer import get_timer_thread
from posthog import redis
from posthog.constants import AnalyticsDBMS
from posthog.internal_metrics import incr, timing
from posthog.settings import (
    CLICKHOUSE_ASYNC,
    CLICKHOUSE_CA,
    CLICKHOUSE_CONN_POOL_MAX,
    CLICKHOUSE_CONN_POOL_MIN,
    CLICKHOUSE_DATABASE,
    CLICKHOUSE_HOST,
    CLICKHOUSE_PASSWORD,
    CLICKHOUSE_SECURE,
    CLICKHOUSE_USER,
    CLICKHOUSE_VERIFY,
    PRIMARY_DB,
    TEST,
)
from posthog.utils import get_safe_cache

InsertParams = Union[list, tuple, types.GeneratorType]
NonInsertParams = Dict[str, Any]
QueryArgs = Optional[Union[InsertParams, NonInsertParams]]

CACHE_TTL = 60  # seconds
SLOW_QUERY_THRESHOLD_MS = 15000
QUERY_TIMEOUT_THREAD = get_timer_thread("ee.clickhouse.client", SLOW_QUERY_THRESHOLD_MS)

_request_information: Optional[Dict] = None


def default_client():
    """
    Return a bare bones client for use in places where we are only interested in general ClickHouse state
    DO NOT USE THIS FOR QUERYING DATA
    """
    return SyncClient(
        host=CLICKHOUSE_HOST,
        secure=CLICKHOUSE_SECURE,
        user=CLICKHOUSE_USER,
        password=CLICKHOUSE_PASSWORD,
        ca_certs=CLICKHOUSE_CA,
        verify=CLICKHOUSE_VERIFY,
    )


def make_ch_pool(**overrides) -> ChPool:
    kwargs = {
        "host": CLICKHOUSE_HOST,
        "database": CLICKHOUSE_DATABASE,
        "secure": CLICKHOUSE_SECURE,
        "user": CLICKHOUSE_USER,
        "password": CLICKHOUSE_PASSWORD,
        "ca_certs": CLICKHOUSE_CA,
        "verify": CLICKHOUSE_VERIFY,
        "connections_min": CLICKHOUSE_CONN_POOL_MIN,
        "connections_max": CLICKHOUSE_CONN_POOL_MAX,
        "settings": {"mutations_sync": "1"} if TEST else {},
        **overrides,
    }

    return ChPool(**kwargs)


if PRIMARY_DB != AnalyticsDBMS.CLICKHOUSE:
    ch_client = None  # type: Client

    class ClickHouseNotConfigured(NotImplementedError):
        def __init__(self, msg='This function only works if PRIMARY_DB is set to indicate ClickHouse!"', *args):
            super().__init__(msg, *args)

    def async_execute(query, args=None, settings=None, with_column_types=False):
        raise ClickHouseNotConfigured()

    def sync_execute(query, args=None, settings=None, with_column_types=False):
        raise ClickHouseNotConfigured()

    def cache_sync_execute(query, args=None, redis_client=None, ttl=None, settings=None, with_column_types=False):
        raise ClickHouseNotConfigured()


else:
    if not TEST and CLICKHOUSE_ASYNC:
        ch_client = Client(
            host=CLICKHOUSE_HOST,
            database=CLICKHOUSE_DATABASE,
            secure=CLICKHOUSE_SECURE,
            user=CLICKHOUSE_USER,
            password=CLICKHOUSE_PASSWORD,
            ca_certs=CLICKHOUSE_CA,
            verify=CLICKHOUSE_VERIFY,
        )

        ch_pool = make_ch_pool()

        @async_to_sync
        async def async_execute(query, args=None, settings=None, with_column_types=False):
            loop = asyncio.get_event_loop()
            task = loop.create_task(
                ch_client.execute(query, args, settings=settings, with_column_types=with_column_types)
            )
            return task

    else:
        # if this is a test use the sync client
        ch_client = SyncClient(
            host=CLICKHOUSE_HOST,
            database=CLICKHOUSE_DATABASE,
            secure=CLICKHOUSE_SECURE,
            user=CLICKHOUSE_USER,
            password=CLICKHOUSE_PASSWORD,
            ca_certs=CLICKHOUSE_CA,
            verify=CLICKHOUSE_VERIFY,
            settings={"mutations_sync": "1"} if TEST else {},
        )

        ch_pool = make_ch_pool()

        def async_execute(query, args=None, settings=None, with_column_types=False):
            return sync_execute(query, args, settings=settings, with_column_types=with_column_types)

    def cache_sync_execute(query, args=None, redis_client=None, ttl=CACHE_TTL, settings=None, with_column_types=False):
        if not redis_client:
            redis_client = redis.get_client()
        key = _key_hash(query, args)
        if redis_client.exists(key):
            result = _deserialize(redis_client.get(key))
            return result
        else:
            result = sync_execute(query, args, settings=settings, with_column_types=with_column_types)
            redis_client.set(key, _serialize(result), ex=ttl)
            return result

    def sync_execute(query, args=None, settings=None, with_column_types=False):
        with ch_pool.get_client() as client:
            start_time = perf_counter()

            prepared_sql, prepared_args, tags = _prepare_query(client=client, query=query, args=args)

            timeout_task = QUERY_TIMEOUT_THREAD.schedule(_notify_of_slow_query_failure, tags)

            try:
                result = client.execute(
                    prepared_sql, params=prepared_args, settings=settings, with_column_types=with_column_types
                )
            except Exception as err:
                err = wrap_query_error(err)
                tags["failed"] = True
                tags["reason"] = type(err).__name__
                incr("clickhouse_sync_execution_failure", tags=tags)

                raise err
            finally:
                execution_time = perf_counter() - start_time

                QUERY_TIMEOUT_THREAD.cancel(timeout_task)
                timing("clickhouse_sync_execution_time", execution_time * 1000.0, tags=tags)

                if app_settings.SHELL_PLUS_PRINT_SQL:
                    print("Execution time: %.6fs" % (execution_time,))
                if _request_information is not None and _request_information.get("save", False):
                    save_query(prepared_sql, execution_time)
        return result

    def substitute_params(query, params):
        """
        Helper method to ease rendering of sql clickhouse queries progressively.
        For example, there are many places where we construct queries to be used
        as subqueries of others. Each time we generate a subquery we also pass
        up the "bound" parameters to be used to render the query, which
        otherwise only happens at the point of calling clickhouse via the
        clickhouse_driver `Client`.

        This results in sometimes large lists of parameters with no relevance to
        the containing query being passed up. Rather than do this, we can
        instead "render" the subqueries prior to using as a subquery, so our
        containing code is only responsible for it's parameters, and we can
        avoid any potential param collisions.
        """
        return cast(SyncClient, ch_client).substitute_params(query, params)


def _prepare_query(client: SyncClient, query: str, args: QueryArgs):
    """
    Given a string query with placeholders we do one of two things:

        1. for a insert query we just format, and remove comments
        2. for non-insert queries, we return the sql with placeholders
        evaluated with the contents of `args`

    We also return `tags` which contains some detail around the context
    within which the query was executed e.g. the django view name

    NOTE: `client.execute` would normally handle substitution, but
    because we want to strip the comments to make it easier to copy
    and past queries from the `system.query_log` easily with metabase
    (metabase doesn't show new lines, so with comments, you can't get
    a working query without exporting to csv or similar), we need to
    do it manually.

    We only want to try to substitue for SELECT queries, which
    clickhouse_driver at this moment in time decides based on the
    below predicate.
    """
    prepared_args: Any = QueryArgs
    if isinstance(args, (list, tuple, types.GeneratorType)):
        # If we get one of these it means we have an insert, let the clickhouse
        # client handle substitution here.
        rendered_sql = query
        prepared_args = args
    elif not args:
        # If `args` is not truthy then make prepared_args `None`, which the
        # clickhouse client uses to signal no substitution is desired. Expected
        # args balue are `None` or `{}` for instance
        rendered_sql = query
        prepared_args = None
    else:
        # Else perform the substitution so we can perform operations on the raw
        # non-templated SQL
        rendered_sql = client.substitute_params(query, args)
        prepared_args = None

    formatted_sql = sqlparse.format(rendered_sql, strip_comments=True)
    annotated_sql, tags = _annotate_tagged_query(formatted_sql, args)

    if app_settings.SHELL_PLUS_PRINT_SQL:
        print()
        print(format_sql(formatted_sql))

    return annotated_sql, prepared_args, tags


def _deserialize(result_bytes: bytes) -> List[Tuple]:
    results = []
    for x in json.loads(result_bytes):
        results.append(tuple(x))
    return results


def _serialize(result: Any) -> bytes:
    return json.dumps(result).encode("utf-8")


def _key_hash(query: str, args: Any) -> bytes:
    key = hashlib.md5(query.encode("utf-8") + json.dumps(args).encode("utf-8")).digest()
    return key


def _annotate_tagged_query(query, args):
    """
    Adds in a /* */ so we can look in clickhouses `system.query_log`
    to easily marry up to the generating code.
    """
    tags = {"kind": (_request_information or {}).get("kind"), "id": (_request_information or {}).get("id")}
    if isinstance(args, dict) and "team_id" in args:
        tags["team_id"] = args["team_id"]
    # Annotate the query with information on the request/task
    if _request_information is not None:
        query = f"/* {_request_information['kind']}:{_request_information['id'].replace('/', '_')} */ {query}"

    return query, tags


def _notify_of_slow_query_failure(tags: Dict[str, Any]):
    tags["failed"] = True
    tags["reason"] = "timeout"
    incr("clickhouse_sync_execution_failure", tags=tags)


def format_sql(rendered_sql, colorize=True):
    formatted_sql = sqlparse.format(rendered_sql, reindent_aligned=True)
    if colorize:
        try:
            import pygments.formatters
            import pygments.lexers

            return pygments.highlight(
                formatted_sql, pygments.lexers.get_lexer_by_name("sql"), pygments.formatters.TerminalFormatter()
            )
        except:
            pass

    return formatted_sql


def save_query(sql: str, execution_time: float) -> None:
    """
    Save query for debugging purposes
    """
    if _request_information is None:
        return

    try:
        key = "save_query_{}".format(_request_information["user_id"])
        queries = json.loads(get_safe_cache(key) or "[]")

        queries.insert(
            0,
            {
                "timestamp": now().isoformat(),
                "query": format_sql(sql, colorize=False),
                "execution_time": execution_time,
            },
        )
        cache.set(key, json.dumps(queries), timeout=120)
    except Exception as e:
        capture_exception(e)
