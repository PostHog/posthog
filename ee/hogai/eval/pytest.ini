[pytest]
python_files = eval_*.py
python_functions = eval_*
env =
    DEBUG=1
    TEST=1
    IN_EVAL_TESTING=1
DJANGO_SETTINGS_MODULE = posthog.settings
addopts = -p no:warnings --reuse-db -s
