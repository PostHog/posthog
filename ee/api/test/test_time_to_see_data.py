import json
from dataclasses import asdict, dataclass, field
from typing import Any, List
from unittest import mock

import pytest

from posthog.client import sync_execute
from posthog.models.query_metrics.sql import (
    CREATE_METRICS_QUERY_LOG,
    CREATE_METRICS_TIME_TO_SEE,
    DROP_METRICS_QUERY_LOG,
    DROP_METRICS_TIME_TO_SEE_TABLE,
)
from posthog.test.base import APIBaseTest


@pytest.mark.usefixtures("unittest_snapshot")
class TestTimeToSeeDataApi(APIBaseTest):
    maxDiff = None
    snapshot: Any

    def setUp(self):
        super().setUp()
        try:
            sync_execute(CREATE_METRICS_TIME_TO_SEE())
            sync_execute(CREATE_METRICS_QUERY_LOG())
        except:
            pass
        self.user.is_staff = True
        self.user.save()

    def tearDown(self):
        super().tearDown()
        sync_execute(DROP_METRICS_TIME_TO_SEE_TABLE())
        sync_execute(DROP_METRICS_QUERY_LOG())

    def test_sessions_api(self):
        insert(
            "metrics_time_to_see_data",
            [
                MetricsRow(
                    session_id="456",
                    timestamp="2022-10-05 12:20:30",
                    time_to_see_data_ms=7000,
                ),
                MetricsRow(
                    session_id="123",
                    timestamp="2022-10-05 10:10:30",
                    time_to_see_data_ms=2000,
                ),
                MetricsRow(
                    session_id="123",
                    timestamp="2022-10-05 10:30:25",
                    time_to_see_data_ms=1000,
                    is_primary_interaction=False,
                ),
                MetricsRow(
                    session_id="123",
                    timestamp="2022-10-05 10:30:30",
                    time_to_see_data_ms=7000,
                ),
            ],
        )

        response = self.client.post("/api/time_to_see_data/sessions").json()
        self.assertEquals(
            response,
            [
                {
                    "duration_ms": 7000,
                    "events_count": 1,
                    "frustrating_interactions_count": 1,
                    "interactions_count": 1,
                    "session_start": "2022-10-05T12:20:23Z",
                    "session_end": "2022-10-05T12:20:30Z",
                    "session_id": "456",
                    "team_events_last_month": mock.ANY,
                    "team_id": 2,
                    "total_interaction_time_to_see_data_ms": 7000,
                    "user": mock.ANY,
                    "user_id": 123,
                },
                {
                    "duration_ms": 1202000,
                    "events_count": 3,
                    "frustrating_interactions_count": 1,
                    "interactions_count": 2,
                    "session_start": "2022-10-05T10:10:28Z",
                    "session_end": "2022-10-05T10:30:30Z",
                    "session_id": "123",
                    "team_events_last_month": mock.ANY,
                    "team_id": 2,
                    "total_interaction_time_to_see_data_ms": 9000,
                    "user": mock.ANY,
                    "user_id": 123,
                },
            ],
        )

    def test_session_events_api(self):
        insert(
            "metrics_time_to_see_data",
            [
                MetricsRow(
                    session_id="456",
                    timestamp="2022-10-05 12:20:30",
                    time_to_see_data_ms=7000,
                ),
                MetricsRow(
                    session_id="123",
                    timestamp="2022-10-05 10:10:30",
                    time_to_see_data_ms=2000,
                    primary_interaction_id="111-222-333",
                ),
                MetricsRow(
                    session_id="123",
                    timestamp="2022-10-05 10:30:25",
                    time_to_see_data_ms=1000,
                    is_primary_interaction=False,
                    primary_interaction_id="111-222-333",
                    query_id="777",
                ),
                MetricsRow(
                    session_id="123",
                    timestamp="2022-10-05 10:30:30",
                    time_to_see_data_ms=7000,
                ),
            ],
        )

        insert(
            "metrics_query_log",
            [
                QueryLogRow(
                    session_id="123",
                    timestamp="2022-10-05 10:10:30",
                    query_duration_ms=400,
                    client_query_id="111-222-333::777",
                ),
                QueryLogRow(
                    session_id="123",
                    timestamp="2022-10-05 10:10:30",
                    query_duration_ms=200,
                    client_query_id="111-222-333::999",
                ),
            ],
        )

        response = self.client.post(
            "/api/time_to_see_data/session_events",
            {
                "team_id": 2,
                "session_start": "2022-10-05T10:10:28Z",
                "session_end": "2022-10-05T10:30:30Z",
                "session_id": "123",
            },
        ).json()

        assert json.dumps(response, indent=4) == self.snapshot


@dataclass
class MetricsRow:
    team_events_last_month: int = 55555
    query_id: str = "123"
    primary_interaction_id: str = "456"
    team_id: int = 2
    user_id: int = 123
    session_id: str = ""
    timestamp: str = "2022-10-05 10:20:00"
    type: str = ""
    context: str = ""
    is_primary_interaction: int = True
    time_to_see_data_ms: int = 200
    status: str = ""
    api_response_bytes: int = 100
    current_url: str = ""
    api_url: str = ""
    insight: str = ""
    action: str = ""
    insights_fetched: int = 1
    insights_fetched_cached: int = 0
    min_last_refresh: str = ""
    max_last_refresh: str = ""


@dataclass
class QueryLogRow:
    host: str = ""
    timestamp: str = ""
    query_duration_ms: int = 555
    read_rows: int = 0
    read_bytes: int = 0
    result_rows: int = 0
    result_bytes: int = 0
    memory_usage: int = 0
    is_initial_query: int = 1
    exception_code: int = 0
    team_id: int = 2
    team_events_last_month: int = 0
    user_id: int = 123
    session_id: str = "123"
    kind: str = ""
    query_type: str = ""
    client_query_id: str = ""
    id: str = ""
    route_id: str = ""
    query_time_range_days: int = 1
    has_joins: int = 0
    has_json_operations: int = 0
    filter_by_type: List[str] = field(default_factory=list)
    breakdown_by: List[str] = field(default_factory=list)
    entity_math: List[str] = field(default_factory=list)
    filter: str = ""
    ProfileEvents: dict = field(default_factory=dict)
    tables: List[str] = field(default_factory=list)
    columns: List[str] = field(default_factory=list)
    query: str = ""
    log_comment = ""


def insert(table: str, rows: List):
    columns = asdict(rows[0]).keys()

    all_values, params = [], {}
    for i, row in enumerate(rows):
        values = ", ".join([f"%(p_{i}_{j})s" for j, _ in enumerate(columns)])
        all_values.append(f"({values})")

        for j, column in enumerate(columns):
            params[f"p_{i}_{j}"] = getattr(row, column)

    sync_execute(
        f"""
        INSERT INTO {table} ({', '.join(columns)})
        VALUES {', '.join(all_values)}
    """,
        params,
    )
