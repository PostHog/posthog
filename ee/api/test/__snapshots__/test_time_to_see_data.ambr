# name: TestTimeToSeeDataApi.test_session_events_api
  '
  {
      "type": "session",
      "data": {
          "session_id": "123",
          "user_id": 123,
          "team_id": 2,
          "session_start": "2022-10-05T10:10:28Z",
          "session_end": "2022-10-05T10:30:30Z",
          "duration_ms": 1202000,
          "team_events_last_month": 55555,
          "events_count": 3,
          "interactions_count": 2,
          "total_interaction_time_to_see_data_ms": 9000,
          "frustrating_interactions_count": 1,
          "user": {
              "distinct_id": "",
              "first_name": "",
              "email": "",
              "is_email_verified": false
          }
      },
      "children": [
          {
              "type": "interaction",
              "data": {
                  "team_events_last_month": 55555,
                  "query_id": "123",
                  "primary_interaction_id": "111-222-333",
                  "team_id": 2,
                  "user_id": 123,
                  "session_id": "123",
                  "timestamp": "2022-10-05T10:10:30",
                  "type": "",
                  "context": "",
                  "is_primary_interaction": 1,
                  "time_to_see_data_ms": 2000,
                  "status": "",
                  "api_response_bytes": 100,
                  "current_url": "",
                  "api_url": "",
                  "insight": "",
                  "action": "",
                  "insights_fetched": 1,
                  "insights_fetched_cached": 0,
                  "min_last_refresh": "1970-01-01T00:00:00",
                  "max_last_refresh": "1970-01-01T00:00:00",
                  "_timestamp": "1970-01-01T00:00:00",
                  "_offset": 0,
                  "_partition": 0,
                  "is_frustrating": 0
              },
              "children": [
                  {
                      "type": "event",
                      "data": {
                          "team_events_last_month": 55555,
                          "query_id": "777",
                          "primary_interaction_id": "111-222-333",
                          "team_id": 2,
                          "user_id": 123,
                          "session_id": "123",
                          "timestamp": "2022-10-05T10:30:25",
                          "type": "",
                          "context": "",
                          "is_primary_interaction": 0,
                          "time_to_see_data_ms": 1000,
                          "status": "",
                          "api_response_bytes": 100,
                          "current_url": "",
                          "api_url": "",
                          "insight": "",
                          "action": "",
                          "insights_fetched": 1,
                          "insights_fetched_cached": 0,
                          "min_last_refresh": "1970-01-01T00:00:00",
                          "max_last_refresh": "1970-01-01T00:00:00",
                          "_timestamp": "1970-01-01T00:00:00",
                          "_offset": 0,
                          "_partition": 0,
                          "is_frustrating": 0
                      },
                      "children": [
                          {
                              "type": "query",
                              "data": {
                                  "host": "",
                                  "timestamp": "2022-10-05T10:10:30",
                                  "query_duration_ms": 400,
                                  "read_rows": 0,
                                  "read_bytes": 0,
                                  "result_rows": 0,
                                  "result_bytes": 0,
                                  "memory_usage": 0,
                                  "is_initial_query": 1,
                                  "exception_code": 0,
                                  "team_id": 2,
                                  "team_events_last_month": 0,
                                  "user_id": 123,
                                  "session_id": "123",
                                  "kind": "",
                                  "query_type": "",
                                  "client_query_id": "111-222-333::777",
                                  "id": "",
                                  "route_id": "",
                                  "query_time_range_days": 1,
                                  "has_joins": 0,
                                  "has_json_operations": 0,
                                  "filter_by_type": [],
                                  "breakdown_by": [],
                                  "entity_math": [],
                                  "filter": "",
                                  "tables": [],
                                  "columns": [],
                                  "query": "",
                                  "log_comment": "",
                                  "is_frustrating": 0
                              },
                              "children": []
                          }
                      ]
                  },
                  {
                      "type": "query",
                      "data": {
                          "host": "",
                          "timestamp": "2022-10-05T10:10:30",
                          "query_duration_ms": 200,
                          "read_rows": 0,
                          "read_bytes": 0,
                          "result_rows": 0,
                          "result_bytes": 0,
                          "memory_usage": 0,
                          "is_initial_query": 1,
                          "exception_code": 0,
                          "team_id": 2,
                          "team_events_last_month": 0,
                          "user_id": 123,
                          "session_id": "123",
                          "kind": "",
                          "query_type": "",
                          "client_query_id": "111-222-333::999",
                          "id": "",
                          "route_id": "",
                          "query_time_range_days": 1,
                          "has_joins": 0,
                          "has_json_operations": 0,
                          "filter_by_type": [],
                          "breakdown_by": [],
                          "entity_math": [],
                          "filter": "",
                          "tables": [],
                          "columns": [],
                          "query": "",
                          "log_comment": "",
                          "is_frustrating": 0
                      },
                      "children": []
                  }
              ]
          },
          {
              "type": "interaction",
              "data": {
                  "team_events_last_month": 55555,
                  "query_id": "123",
                  "primary_interaction_id": "456",
                  "team_id": 2,
                  "user_id": 123,
                  "session_id": "123",
                  "timestamp": "2022-10-05T10:30:30",
                  "type": "",
                  "context": "",
                  "is_primary_interaction": 1,
                  "time_to_see_data_ms": 7000,
                  "status": "",
                  "api_response_bytes": 100,
                  "current_url": "",
                  "api_url": "",
                  "insight": "",
                  "action": "",
                  "insights_fetched": 1,
                  "insights_fetched_cached": 0,
                  "min_last_refresh": "1970-01-01T00:00:00",
                  "max_last_refresh": "1970-01-01T00:00:00",
                  "_timestamp": "1970-01-01T00:00:00",
                  "_offset": 0,
                  "_partition": 0,
                  "is_frustrating": 1
              },
              "children": []
          }
      ]
  }
  '
---
